File: react-ios/src/services/workoutExportService.ts
import { Paths, File } from 'expo-file-system';
import Constants from 'expo-constants';
import { getCompletedSessions } from '@/db/sessionRepository';
import { getAllWorkoutPlans } from '@/db/repository';
import { getDatabase } from '@/db/index';
import type { WorkoutSession, WorkoutPlan, PlannedExercise, PlannedSet, SessionExercise, SessionSet, GymRow, UserSettingsRow } from '@/types';

/**
 * Export completed workout sessions as a portable JSON file.
 * Strips internal IDs to produce a clean, shareable format.
 * Returns the file URI for sharing.
 */
export async function exportSessionsAsJson(): Promise<string> {
  const sessions = await getCompletedSessions();

  if (sessions.length === 0) {
    throw new ExportError('No completed workouts to export.');
  }

  const exportData = {
    exportedAt: new Date().toISOString(),
    appVersion: Constants.expoConfig?.version || '1.0.0',
    sessions: sessions.map(stripSession),
  };

  const timestamp = new Date()
    .toISOString()
    .replace(/:/g, '-')
    .replace(/\.\d{3}Z$/, '')
    .replace('T', '_');
  const fileName = `liftmark_workouts_${timestamp}.json`;
  const exportFile = new File(Paths.cache, fileName);

  exportFile.write(JSON.stringify(exportData, null, 2));

  return exportFile.uri;
}

/**
 * Export a single workout session as a portable JSON file.
 * Returns the file URI for sharing.
 */
export async function exportSingleSessionAsJson(session: WorkoutSession): Promise<string> {
  const exportData = {
    exportedAt: new Date().toISOString(),
    appVersion: Constants.expoConfig?.version || '1.0.0',
    session: stripSession(session),
  };

  const fileName = buildSessionFileName(session.name, session.date);
  const exportFile = new File(Paths.cache, fileName);

  exportFile.write(JSON.stringify(exportData, null, 2));

  return exportFile.uri;
}

/**
 * Build a sanitized file name: workout-{name}-{date}.json
 */
export function buildSessionFileName(name: string, date: string): string {
  const sanitized = name
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')     // strip diacritics
    .replace(/[^\w\s-]/g, '')            // remove non-alphanumeric (except spaces/hyphens)
    .replace(/\s+/g, '-')               // spaces to hyphens
    .replace(/-+/g, '-')                // collapse multiple hyphens
    .replace(/^-|-$/g, '')              // trim leading/trailing hyphens
    .slice(0, 50);

  const datePart = date.split('T')[0] || new Date().toISOString().split('T')[0];
  const namePart = sanitized || 'workout';

  return `workout-${namePart}-${datePart}.json`;
}

export class ExportError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ExportError';
  }
}

/**
 * Export all app data as a unified JSON file for backup/transfer.
 * Includes plans, sessions, gyms, and safe settings (no API keys).
 * Returns the file URI for sharing.
 */
export async function exportUnifiedJson(): Promise<string> {
  const plans = await getAllWorkoutPlans();
  const sessions = await getCompletedSessions();
  const db = await getDatabase();

  // Read gyms
  const gymRows = await db.getAllAsync<GymRow>('SELECT * FROM gyms');
  const gyms = gymRows.map(row => {
    const gym: Record<string, unknown> = {
      name: row.name,
      isDefault: row.is_default === 1,
    };
    if (row.created_at) gym.createdAt = row.created_at;
    return gym;
  });

  // Read settings (exclude sensitive data like API keys)
  const settingsRow = await db.getFirstAsync<UserSettingsRow>(
    'SELECT * FROM user_settings LIMIT 1'
  );
  const settings: Record<string, unknown> = {};
  if (settingsRow) {
    settings.defaultWeightUnit = settingsRow.default_weight_unit;
    settings.enableWorkoutTimer = settingsRow.enable_workout_timer === 1;
    settings.autoStartRestTimer = settingsRow.auto_start_rest_timer === 1;
    settings.theme = settingsRow.theme;
    settings.keepScreenAwake = settingsRow.keep_screen_awake === 1;
    if (settingsRow.custom_prompt_addition) {
      settings.customPromptAddition = settingsRow.custom_prompt_addition;
    }
  }

  const exportData = {
    formatVersion: '1.0',
    exportedAt: new Date().toISOString(),
    appVersion: Constants.expoConfig?.version || '1.0.0',
    plans: plans.map(stripPlan),
    sessions: sessions.map(stripSession),
    gyms,
    settings,
  };

  const timestamp = new Date()
    .toISOString()
    .replace(/:/g, '-')
    .replace(/\.\d{3}Z$/, '')
    .replace('T', '_');
  const fileName = `liftmark_export_${timestamp}.json`;
  const exportFile = new File(Paths.cache, fileName);

  exportFile.write(JSON.stringify(exportData, null, 2));

  return exportFile.uri;
}

function stripPlan(plan: WorkoutPlan) {
  const result: Record<string, unknown> = {
    name: plan.name,
    exercises: plan.exercises
      .filter(e => e.sets.length > 0 || e.groupType != null)
      .map(stripPlannedExercise),
  };
  if (plan.description) result.description = plan.description;
  if (plan.tags.length > 0) result.tags = plan.tags;
  if (plan.defaultWeightUnit) result.defaultWeightUnit = plan.defaultWeightUnit;
  if (plan.sourceMarkdown) result.sourceMarkdown = plan.sourceMarkdown;
  result.isFavorite = plan.isFavorite ?? false;
  return result;
}

function stripPlannedExercise(exercise: PlannedExercise) {
  const result: Record<string, unknown> = {
    exerciseName: exercise.exerciseName,
    orderIndex: exercise.orderIndex,
    sets: exercise.sets.map(stripPlannedSet),
  };
  if (exercise.notes) result.notes = exercise.notes;
  if (exercise.equipmentType) result.equipmentType = exercise.equipmentType;
  if (exercise.groupType) result.groupType = exercise.groupType;
  if (exercise.groupName) result.groupName = exercise.groupName;
  return result;
}

function stripPlannedSet(set: PlannedSet) {
  const result: Record<string, unknown> = {
    orderIndex: set.orderIndex,
    isDropset: set.isDropset ?? false,
    isPerSide: set.isPerSide ?? false,
  };
  if (set.targetWeight != null) result.targetWeight = set.targetWeight;
  if (set.targetWeightUnit) result.targetWeightUnit = set.targetWeightUnit;
  if (set.targetReps != null) result.targetReps = set.targetReps;
  if (set.targetTime != null) result.targetTime = set.targetTime;
  if (set.targetRpe != null) result.targetRpe = set.targetRpe;
  if (set.restSeconds != null) result.restSeconds = set.restSeconds;
  if (set.tempo) result.tempo = set.tempo;
  if (set.notes) result.notes = set.notes;
  return result;
}

function stripSession(session: WorkoutSession) {
  return {
    name: session.name,
    date: session.date,
    startTime: session.startTime,
    endTime: session.endTime,
    duration: session.duration,
    notes: session.notes,
    status: session.status,
    exercises: session.exercises.map(stripExercise),
  };
}

function stripExercise(exercise: SessionExercise) {
  return {
    exerciseName: exercise.exerciseName,
    orderIndex: exercise.orderIndex,
    notes: exercise.notes,
    equipmentType: exercise.equipmentType,
    groupType: exercise.groupType,
    groupName: exercise.groupName,
    status: exercise.status,
    sets: exercise.sets.map(stripSet),
  };
}

function stripSet(set: SessionSet) {
  return {
    orderIndex: set.orderIndex,
    targetWeight: set.targetWeight,
    targetWeightUnit: set.targetWeightUnit,
    targetReps: set.targetReps,
    targetTime: set.targetTime,
    targetRpe: set.targetRpe,
    restSeconds: set.restSeconds,
    actualWeight: set.actualWeight,
    actualWeightUnit: set.actualWeightUnit,
    actualReps: set.actualReps,
    actualTime: set.actualTime,
    actualRpe: set.actualRpe,
    completedAt: set.completedAt,
    status: set.status,
    notes: set.notes,
    tempo: set.tempo,
    isDropset: set.isDropset,
    isPerSide: set.isPerSide,
  };
}

File: mobile-apps/ios/LiftMark/Services/WorkoutExportService.swift
import Foundation
import GRDB

/// Errors thrown during workout export operations.
enum ExportError: LocalizedError {
    case noCompletedWorkouts
    case fileWriteFailed(String)

    var errorDescription: String? {
        switch self {
        case .noCompletedWorkouts:
            return "No completed workouts to export."
        case .fileWriteFailed(let reason):
            return "Failed to write export file: \(reason)"
        }
    }
}

/// Service for exporting workout sessions as portable JSON files.
/// Strips internal database IDs to produce clean, shareable data.
struct WorkoutExportService {
    private let repository = SessionRepository()

    /// Export all completed sessions to a single JSON file.
    /// Returns the file URL for sharing.
    func exportSessionsAsJson() throws -> URL {
        let sessions = try repository.getCompleted()

        guard !sessions.isEmpty else {
            throw ExportError.noCompletedWorkouts
        }

        let exportData: [String: Any] = [
            "exportedAt": ISO8601DateFormatter().string(from: Date()),
            "appVersion": appVersion(),
            "sessions": sessions.map { stripSession($0) }
        ]

        let timestamp = ISO8601DateFormatter().string(from: Date())
            .replacingOccurrences(of: ":", with: "-")
            .replacingOccurrences(of: "T", with: "_")
            .replacingOccurrences(of: "Z", with: "")
        let fileName = "liftmark_workouts_\(timestamp).json"

        return try writeExportFile(exportData, fileName: fileName)
    }

    /// Export a single workout session as a portable JSON file.
    /// Returns the file URL for sharing.
    func exportSingleSessionAsJson(_ session: WorkoutSession) throws -> URL {
        let exportData: [String: Any] = [
            "exportedAt": ISO8601DateFormatter().string(from: Date()),
            "appVersion": appVersion(),
            "session": stripSession(session)
        ]

        let fileName = buildSessionFileName(name: session.name, date: session.date)
        return try writeExportFile(exportData, fileName: fileName)
    }

    /// Build a sanitized file name: workout-{name}-{date}.json
    func buildSessionFileName(name: String, date: String) -> String {
        var sanitized = name.lowercased()
        sanitized = sanitized.folding(options: .diacriticInsensitive, locale: .current)
        sanitized = sanitized.replacingOccurrences(
            of: "[^a-z0-9\\s-]",
            with: "",
            options: .regularExpression
        )
        sanitized = sanitized.replacingOccurrences(
            of: "\\s+",
            with: "-",
            options: .regularExpression
        )
        sanitized = sanitized.replacingOccurrences(
            of: "-+",
            with: "-",
            options: .regularExpression
        )
        sanitized = sanitized.trimmingCharacters(in: CharacterSet(charactersIn: "-"))
        if sanitized.count > 50 {
            sanitized = String(sanitized.prefix(50))
        }

        let datePart = date.split(separator: "T").first.map(String.init)
            ?? ISO8601DateFormatter().string(from: Date()).split(separator: "T").first.map(String.init)
            ?? "unknown"
        let namePart = sanitized.isEmpty ? "workout" : sanitized

        return "workout-\(namePart)-\(datePart).json"
    }

    /// Export all app data as a unified JSON file for backup/transfer.
    /// Includes plans, sessions, gyms, and safe settings (no API keys).
    func exportUnifiedJson() throws -> URL {
        let planRepo = WorkoutPlanRepository()
        let plans = try planRepo.getAll()
        let sessions = try repository.getCompleted()

        // Read gyms directly from database
        let gyms: [[String: Any]] = (try? DatabaseManager.shared.database().read { db -> [[String: Any]] in
            let tableExists = try Bool.fetchOne(db, sql: "SELECT COUNT(*) > 0 FROM sqlite_master WHERE type='table' AND name='gyms'") ?? false
            guard tableExists else { return [] }

            let rows = try Row.fetchAll(db, sql: "SELECT * FROM gyms")
            return rows.map { row in
                var dict: [String: Any] = [
                    "name": (row["name"] as? String) ?? "",
                    "isDefault": ((row["is_default"] as? Int) ?? 0) == 1
                ]
                if let createdAt = row["created_at"] as? String { dict["createdAt"] = createdAt }
                return dict
            }
        }) ?? []

        // Read settings (strip sensitive data)
        let settings: [String: Any] = (try? DatabaseManager.shared.database().read { db in
            // Check if settings table exists
            let tableExists = try Bool.fetchOne(db, sql: "SELECT COUNT(*) > 0 FROM sqlite_master WHERE type='table' AND name='settings'") ?? false
            guard tableExists else { return [:] as [String: Any] }

            guard let row = try Row.fetchOne(db, sql: "SELECT * FROM settings LIMIT 1") else {
                return [:] as [String: Any]
            }
            var dict: [String: Any] = [:]
            if let v = row["default_weight_unit"] as? String { dict["defaultWeightUnit"] = v }
            if let v = row["enable_workout_timer"] as? Int { dict["enableWorkoutTimer"] = v == 1 }
            if let v = row["auto_start_rest_timer"] as? Int { dict["autoStartRestTimer"] = v == 1 }
            if let v = row["theme"] as? String { dict["theme"] = v }
            if let v = row["keep_screen_awake"] as? Int { dict["keepScreenAwake"] = v == 1 }
            if let v = row["custom_prompt_addition"] as? String { dict["customPromptAddition"] = v }
            return dict
        }) ?? [:]

        let exportData: [String: Any] = [
            "formatVersion": "1.0",
            "exportedAt": ISO8601DateFormatter().string(from: Date()),
            "appVersion": appVersion(),
            "plans": plans.map { stripPlan($0) },
            "sessions": sessions.map { stripSession($0) },
            "gyms": gyms,
            "settings": settings
        ]

        let timestamp = ISO8601DateFormatter().string(from: Date())
            .replacingOccurrences(of: ":", with: "-")
            .replacingOccurrences(of: "T", with: "_")
            .replacingOccurrences(of: "Z", with: "")
        let fileName = "liftmark_export_\(timestamp).json"

        return try writeExportFile(exportData, fileName: fileName)
    }

    // MARK: - Private

    private func appVersion() -> String {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    }

    private func writeExportFile(_ data: [String: Any], fileName: String) throws -> URL {
        let jsonData = try JSONSerialization.data(withJSONObject: data, options: [.prettyPrinted, .sortedKeys])
        guard let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
            throw ExportError.fileWriteFailed("Cache directory is unavailable")
        }
        let fileURL = cacheDir.appendingPathComponent(fileName)
        try jsonData.write(to: fileURL)
        return fileURL
    }

    private func stripPlan(_ plan: WorkoutPlan) -> [String: Any] {
        var dict: [String: Any] = [
            "name": plan.name,
            "exercises": plan.exercises.filter { !$0.sets.isEmpty || $0.groupType != nil }.map { stripPlannedExercise($0) }
        ]
        if let desc = plan.description { dict["description"] = desc }
        if !plan.tags.isEmpty { dict["tags"] = plan.tags }
        if let unit = plan.defaultWeightUnit { dict["defaultWeightUnit"] = unit.rawValue }
        if let md = plan.sourceMarkdown { dict["sourceMarkdown"] = md }
        dict["isFavorite"] = plan.isFavorite
        return dict
    }

    private func stripPlannedExercise(_ exercise: PlannedExercise) -> [String: Any] {
        var dict: [String: Any] = [
            "exerciseName": exercise.exerciseName,
            "orderIndex": exercise.orderIndex,
            "sets": exercise.sets.map { stripPlannedSet($0) }
        ]
        if let notes = exercise.notes { dict["notes"] = notes }
        if let equip = exercise.equipmentType { dict["equipmentType"] = equip }
        if let gt = exercise.groupType { dict["groupType"] = gt.rawValue }
        if let gn = exercise.groupName { dict["groupName"] = gn }
        return dict
    }

    private func stripPlannedSet(_ set: PlannedSet) -> [String: Any] {
        var dict: [String: Any] = [
            "orderIndex": set.orderIndex,
            "isDropset": set.isDropset,
            "isPerSide": set.isPerSide,
            "isAmrap": set.isAmrap
        ]
        if let v = set.targetWeight { dict["targetWeight"] = v }
        if let v = set.targetWeightUnit { dict["targetWeightUnit"] = v.rawValue }
        if let v = set.targetReps { dict["targetReps"] = v }
        if let v = set.targetTime { dict["targetTime"] = v }
        if let v = set.targetRpe { dict["targetRpe"] = v }
        if let v = set.restSeconds { dict["restSeconds"] = v }
        if let v = set.tempo { dict["tempo"] = v }
        if let v = set.notes { dict["notes"] = v }
        return dict
    }

    private func stripSession(_ session: WorkoutSession) -> [String: Any] {
        var dict: [String: Any] = [
            "name": session.name,
            "date": session.date,
            "status": session.status.rawValue,
            "exercises": session.exercises.map { stripExercise($0) }
        ]
        if let startTime = session.startTime { dict["startTime"] = startTime }
        if let endTime = session.endTime { dict["endTime"] = endTime }
        if let duration = session.duration { dict["duration"] = duration }
        if let notes = session.notes { dict["notes"] = notes }
        return dict
    }

    private func stripExercise(_ exercise: SessionExercise) -> [String: Any] {
        var dict: [String: Any] = [
            "exerciseName": exercise.exerciseName,
            "orderIndex": exercise.orderIndex,
            "status": exercise.status.rawValue,
            "sets": exercise.sets.map { stripSet($0) }
        ]
        if let notes = exercise.notes { dict["notes"] = notes }
        if let equipmentType = exercise.equipmentType { dict["equipmentType"] = equipmentType }
        if let groupType = exercise.groupType { dict["groupType"] = groupType.rawValue }
        if let groupName = exercise.groupName { dict["groupName"] = groupName }
        return dict
    }

    private func stripSet(_ set: SessionSet) -> [String: Any] {
        var dict: [String: Any] = [
            "orderIndex": set.orderIndex,
            "status": set.status.rawValue,
            "isDropset": set.isDropset,
            "isPerSide": set.isPerSide
        ]
        if let v = set.targetWeight { dict["targetWeight"] = v }
        if let v = set.targetWeightUnit { dict["targetWeightUnit"] = v.rawValue }
        if let v = set.targetReps { dict["targetReps"] = v }
        if let v = set.targetTime { dict["targetTime"] = v }
        if let v = set.targetRpe { dict["targetRpe"] = v }
        if let v = set.restSeconds { dict["restSeconds"] = v }
        if let v = set.actualWeight { dict["actualWeight"] = v }
        if let v = set.actualWeightUnit { dict["actualWeightUnit"] = v.rawValue }
        if let v = set.actualReps { dict["actualReps"] = v }
        if let v = set.actualTime { dict["actualTime"] = v }
        if let v = set.actualRpe { dict["actualRpe"] = v }
        if let v = set.completedAt { dict["completedAt"] = v }
        if let v = set.notes { dict["notes"] = v }
        if let v = set.tempo { dict["tempo"] = v }
        return dict
    }
}
